src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs

```diff
@@ -15,45 +15,50 @@ public partial class Solution
     {
         private partial class CompilationTracker
         {
-            // An empty compilation 
-            // Used as a base class to hold the SkeletonReference property used by derived types.
+            /// <summary>
+            /// The base type of all <see cref="CompilationTracker"/> states. The state of a <see cref="CompilationTracker" />
+            /// starts at <see cref="Empty"/>, and then will progress through the other states until it finally reaches
+            /// <see cref="FinalState" />.
+            /// </summary>
             private class State
             {
-                public static readonly State Empty = new State();
-
-                // strong reference to declaration only compilation. 
-                // this doesn't make any expensive information such as symbols or references alive. just
-                // things like declaration table alive.
+                /// <summary>
+                /// The base <see cref="State"/> that starts with everything empty.
+                /// </summary>
+                public static readonly State Empty = new State(compilation: ConstantValueSource<Compilation>.Empty, declarationOnlyCompilation: null);
+
+                /// <summary>
+                /// A strong reference to the declaration-only compilation. This compilation isn't used to produce symbols,
+                /// nor does it have any references. It just holds the declaration table alive.
+                /// </summary>
                 public Compilation DeclarationOnlyCompilation { get; }
 
-                // The compilation available.  May be an InProgress, Full Declaration, or Final compilation
+                /// <summary>
+                /// The best compilation that is available.  May be an in-progress, full declaration, or a final compilation.
+                /// </summary>
                 public ValueSource<Compilation> Compilation { get; }
 
-                // The Final compilation if available, otherwise an empty IValueSource
+                /// <summary>
+                /// Specifies if there are references that got dropped in the production of <see cref="FinalCompilation"/>. This can return
+                /// null if the state isn't at the point where it would know, and it's necessary to transition to <see cref="FinalState"/> to figure that out.
+                /// </summary>
+                public virtual bool? HasIncompleteReferences => null;
```

Comment:
This is how this pattern is done in the compilation tracker. If you call `GetCompilation()` it first checks the state and sees if it already has a return value for `FinalCompilation` below, and returns it. Otherwise it goes through it's process to bring it to the final state. So this is _consistent_ with the pattern. I agree it's somewhat strange though, but we could try moving both of them down.